### ISSUE TITLE
`opa fmt` changes `\u0000` to `\x00` which leads to parser error

### ISSUE THREAD

**Issue by zregvart** [open-policy-agent/opa#6220]
## Short description 

`opa fmt` will reformat this Rego ([playground](https://play.openpolicyagent.org/p/O4TLbhivHd)):

```rego
package play

x := "\u0000"

```

as

```rego
package play

x := "\x00"

```

Which causes `rego_parse_error: illegal escape sequence`.

## Steps To Reproduce
1. run `opa eval 'nul := "\u0000"'` see the result and `nul` bound to `"\u0000"`
2. run `echo 'package test\nnul := "\\u0000"' | opa fmt` to see the conversion to `nul := "\x00"`
3. opa eval 'nul := "\x00"' to see the `rego_parse_error: illegal escape sequence` error (in JSON)

## Expected behavior

Formatting the code using `opa fmt` should not introduce parser errors. Not sure if `opa fmt` needs to be changed or the parser needs to allow `\x...` escape sequences.

## Additional context

```
$ opa version
Version: 0.53.1
Build Commit: 947d65b372eebfde65686028a0c43b30dd7f04b2
Build Timestamp: 2023-06-06T07:12:49Z
Build Hostname: e0cbbd3ede6c
Go Version: go1.20.4
Platform: linux/amd64
WebAssembly: available
```

**Comment by anderseknert** [comment]
Hi @zregvart, and thanks for reporting this 👍 

Looks like the formatter calls the `String()` method on the term, which delegates to [strconv.Quote](https://github.com/open-policy-agent/opa/blob/main/ast/term.go#L801) and that's where the string is transformed into the "illegal" entity `\x00`. I'm honestly not sure about what the correct thing to do would be. I can't discern from the docs whether control characters in strings are supposed to be allowed or not 😅 So I guess whatever the outcome here is, it'd be good to add to the [docs on strings](https://www.openpolicyagent.org/docs/latest/policy-language/#strings) what is supported and what isn't. And if they aren't allowed, the formatter should obviously be fixed to flag them.

**Comment by zregvart** [comment]
Just noting my workaround which is to use `nul := base64.decode("AA==")`. I don't think many folk will run into this issue. I needed it for calculating a digest of a specific format that has `\0` as delimiter.

**Comment by anderseknert** [comment]
Yep, I think any character that's valid JSON should be allowed, and as far as I know, using `\u0000` to represent unicode null is valid, so the formatter should leave it be... but perhaps there are reasons for this behavior I'm not aware of, and somebody else could chip in to provide some background on that.

**Comment by MaxymVlasov** [comment]
Get the same issue during writing OPA tests for Spacelift Notification policy.

Here is minimized version of ~200 lines input
```rego
package spacelift
import rego.v1

input_not_in_cell := {
	"account": {
		"name": "qwerty"
	},
	# Minimized from "echo '\n\n\u001b[0m\u001b[32mInit tflint\u001b[0m'" which is used to colorize output in Spacelift UI
	"before_init": [ "echo '\u001b[0m'" ], 
}
```
https://play.openpolicyagent.org/p/20T7b7fg9Y

and it is still able to show that `opa fmt` just crash on one of first entities in object, not where actually it detects error, which make it hard to debug reasons why it crashed

```rego
notification.webhook_test.rego was successfully formatted, but the result is invalid: 2 errors occurred:
formatted:12: rego_parse_error: unexpected { token: non-terminated object
        "account": {"name": "qwerty"},
                   ^
formatted:12: rego_parse_error: unexpected { token: expected rule value term (e.g., input_not_in_cell := <VALUE> { ... })
        "account": {"name": "qwerty"},
                   ^

To inspect the formatted Rego, you can turn off this check with --check-result=false.
```

**Comment by anderseknert** [comment]
Would it work using a raw string for your purpose, @MaxymVlasov? i.e. instead of

```
"echo '\u001b[0m'"
```
->
```
`echo '\u001b[0m'`
```

**Comment by MaxymVlasov** [comment]
No, because it still means modification of inputs provided by vendor - https://docs.spacelift.io/concepts/policy/notification-policy#data-input

And any modifications mean that firstly you need to identify where the issue is, as the error message is not helpful at all.
For my specific case, I can simply drop this line as we do not use data from it, but firstly I spent about half-day trying to understand what's could be wrong to find the root cause.

**Comment by anderseknert** [comment]
Sure, that's [a bug](https://github.com/open-policy-agent/opa/issues/4263) in its own right. And so is the formatter handling this issue describes. I don't have much experience working with Spacelift, which is probably why I don't fully get what you're saying about the input data they provide. How's that input data connected to the policy from your example? They have provided a policy that can't be formatted?

**Comment by MaxymVlasov** [comment]
When you run your stack in CI (Spacelift in this case), vendor provide you some info related to the current run as inputs, so you can build some custom logic to autoapprove certain changes, track some metrics and so on.

Best way to test your policy - is getting a few inputs "as is" from different stacks, and write tests to it. If it's a simple policy, you could just remove >95% of such inputs just to save space and not violate `regal file-length` check as you don't use that info, but for complex scenarios better leave inputs as is.

They're how you can get inputs for local `opa test`:
1. Each run can have a bunch of policies with different inputs:
![Screenshot from 2024-12-12 15-30-52](https://github.com/user-attachments/assets/1d42e551-070b-44e8-a58d-6a26a5d07432)
2. To get them, just need to expand policy execution and copy `Input` to related variable in `_test.rego` file
![Screenshot from 2024-12-12 15-31-42](https://github.com/user-attachments/assets/6e6c84d9-06d7-49c7-99f2-0177c619ee07)
3. Then, having the same inputs as Spacelift provided to policy during policy execution, we can actually write tests to our policy

**Comment by anderseknert** [comment]
Thanks @MaxymVlasov! That's very helpful. 

So you're taking the input from Spacelift and paste that into a Rego test on your own premises, where you can test your policy standalone? And naturally you want to be able to format and lint that policy as any other, but the formatter strips some of the strings, making the policy unparseable. 

I'll try to set some time aside to fix this. In the meantime, you may be able to work around the issue by either:

1. Use raw strings (ie wrapped in backticks) for all string data used for input
2. Move the data out of your policies and into .json files that you read as part of running the tests

Let me know if none of those options work, and we need to figure something else out.

**Comment by MaxymVlasov** [comment]
Yep, you described exactly what I do

p.2 can work, but how I can include external files then? I can't find quickly any docs for it
IE, I have 
```
myinput1.json
myinput2.json
policy.rego
poily_test.rego
```

How I can do in `policy_test.rego` something like:

```rego
input1 := read_file("myinput1.json")
input2 := read_file("myinput2.json")
```
I see [rego.parse_module](https://www.openpolicyagent.org/docs/latest/policy-reference/#builtin-rego-regoparse_module) which in theory could do this trick, but I not sure how to use it. Will much apreciate any example

**Comment by anderseknert** [comment]
If you run something like `opa test .` , the test runner will read both Rego and JSON files from the directory (and subdirectories). Note that all the JSON data will be merged into a single document, so you may want to "namespace" the data per test to avoid conflicts in that merge.. i.e. have your files look something like:

```json
{
  "testdata": {
    "test1": {}
  }
}
```

WIth a structure like that, your tests can then reference `data.testdata.test1` and so on. 

You can either put the data in one big file, or have as many JSON files as you wish. The only thing that's important is the namespacing, so that each piece of data is under its own unique path.
